Feature/#194 add json writer

proohit opened the pull request:
### Description

Linked issue: #194 

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Code style follows the tinylog standard
- [x] All classes and methods have Javadoc
- [x] Changes are covered by JUnit tests including corner cases, errors, and exception handling
- [x] Maven build works including compiling, tests, and checks (`mvn verify`)
- [x] Changes are committed by a verified email address that is assigned to the GitHub account (https://github.com/settings/emails)

### Documentation

Additions or amendments for the [public documentation](https://github.com/pmwmedia/tinylog/wiki/Documentation):


Writer                                      | Property Value | Description
:------------------------------------------ |:-------------- |:------------
...
[JSON Writer](#json-writer)                 | json           | Writes log entries to a json file

...
### JSON Writer

The JSON writer writes log entries to a defined json file. It supports custom field configuration with the syntax `writer.field.FIELDNAME = FORMAT`. With appending mode off, the file is reset with new log entries. The format can be configured like in other writers.
Example:

```properties
writer         = json
writer.file    = log.json
writer.charset = UTF-8 
writer.append  = true
writer.field.message = message
writer.field.date = date
writer.field.lvl = level
```


### Agreements

- [x] I agree that my changes will be published under the terms of the [Apache License 2.0](https://github.com/pmwmedia/tinylog/blob/v2.0/license.txt)
- [x] I agree that my GitHub user name will be published in the release notes

proohit committed:
Added JSONWriter implementation first attempt

proohit committed:
#194 - Added Unit tests for JSONWriter

proohit committed:
- #194 - Add threadToken

proohit committed:
- #194 - Refactoring and cleanup

proohit committed:
#194 - Rename to match checkstyle

proohit committed:
#194 - Fix checkstyle and spotbugs

proohit committed:
#194 - return defaultCharset

proohit committed:
#194 - organize imports

proohit commented:
Whoaw. I updated the wiki page and thought it would add it to the PR but it instantly saved it. Sorry for that, maybe we can rollback the page until this PR is merged.

proohit committed:
#194 - Added Writer to services

proohit committed:
#194 - Fixed wrong line seperator

proohit commented:
I discovered some problems with the initial appending logic, thus I converted this PR to draft.

Maybe someone has an idea? The current solution expects the "]" to be the last character (including newlines). Truncating the bytes of "]" may or may not delete the "]" character depending on whether it's the last character or not.

I wanted to read the stream until I find bytes that match those of "]" and truncate all bytes after it but the stream we use is an OutputStream, so it has no API to read bytes. This needs further investigation

proohit committed:
#194 - Fixed appending bug

proohit committed:
#194 - Added Javadoc

proohit committed:
#194 - Remove try-with resources

pmwmedia commented:
Thank your for your awesome, pull request! Your implementation looks very solid and promising :) According to your commits, I see that you found out all tinylog's build specifics by yourself. I know that it isn't such easy on the first attempt as tinylog 2 still supports old Java versions including Android 1. Thank you very much that you also wrote good Javadoc documentation and JUnit tests.

> Whoaw. I updated the wiki page and thought it would add it to the PR but it instantly saved it. Sorry for that, maybe we can rollback the page until this PR is merged.

Unfortunately, GitHub doesn't support pull requests for the Wiki repo. It was also surprised when I initially set the Wiki up. However, it is very easy to revert a commit via Git.

pmwmedia commented on configuration/spotbugs-filters.xml:
```diff
@@ -42,6 +42,12 @@
+		<Class name="org.tinylog.writers.JsonWriter" />
+		<!-- It doesn't matter if "File.mkdirs()" creates new directories (returns "true") or if these exist already (returns "false") -->
+		<Bug pattern="RV_RETURN_VALUE_IGNORED_BAD_PRACTICE" />
+	</Match>
```
Could you merge the JsonWriter and AbstractFormatPatternWriter match into a single match? They both disable the same bug pattern. You can combine both classes by an OR clause.

proohit replied:
Fixed in https://github.com/tinylog-org/tinylog/pull/195/commits/6aa7cc8f70f0b8cb16a52beb9b4459c3348d37b8

pmwmedia commented on tinylog-impl/src/main/java/org/tinylog/writers/JsonWriter.java:
```diff
@@ -0,0 +1,268 @@
+	 * @throws IllegalArgumentException Log file is not defined in configuration
+	 */
+	public JsonWriter(final Map<String, String> properties) throws IOException {
+
```
The redundant empty line can be removed (minor code style violation).

proohit replied:
Fixed in https://github.com/tinylog-org/tinylog/pull/195/commits/5322e7bbaaa6040dfad72f601cac39c19db08998

pmwmedia commented on tinylog-impl/src/main/java/org/tinylog/writers/JsonWriter.java:
```diff
@@ -0,0 +1,268 @@
+			return charsetName == null ? Charset.defaultCharset() : Charset.forName(charsetName);
+		} catch (IllegalArgumentException ex) {
+			InternalLogger.log(Level.ERROR, "Invalid charset: " + charsetName);
+			return Charset.defaultCharset();
```
Could you add a JUnit test for the error case?

proohit replied:
Fixed in https://github.com/tinylog-org/tinylog/pull/195/commits/a849524287ae1533012edf7851b32d03b7e289ef

pmwmedia commented on tinylog-impl/src/main/java/org/tinylog/writers/JsonWriter.java:
```diff
@@ -0,0 +1,268 @@
+	private void writeProperty(final Token token, final LogEntry logEntry, final StringBuilder builder,
+			final boolean hasPrevious) {
+		if (hasPrevious) {
+			builder.append(String.format(",%s", NEW_LINE));
```
String format is quite slow. You could improve the performance with `builder.append(',').append(NEW_LINE);`

proohit replied:
Fixed in https://github.com/tinylog-org/tinylog/pull/195/commits/5322e7bbaaa6040dfad72f601cac39c19db08998

pmwmedia commented on tinylog-impl/src/main/java/org/tinylog/writers/JsonWriter.java:
```diff
@@ -0,0 +1,268 @@
+
+	@Override
+	public Collection<LogEntryValue> getRequiredLogEntryValues() {
+		return EnumSet.noneOf(LogEntryValue.class);
```
Here, tinylog expects all required log entry values from all used tokens, created via the format pattern parsers. You can just create a new enum set, iterate over all tokens, and add their required log entry values to the enum set. If you return, an empty enum set, it is not guaranteed that further tinylog version won't pass only empty log entries to the JSON parser as optimization result.

proohit replied:
Fixed in https://github.com/tinylog-org/tinylog/pull/195/commits/5322e7bbaaa6040dfad72f601cac39c19db08998

proohit commented:
Hi, no probs there as I use this logger in one of my projects :). If you're curious, here's a link: [DirisSlave discord bot](https://github.com/proohit/DirisSlave)

I figured out most things like checkstyle, spotbugs etc with your requirement on `mvn verify`. I think we should definitely add a CONTRIBUTING.md guideline. I had some problems with other packages like tinylog-api, but I ignored them as I didn't change anything there.

Also I didn't write many tests. I'm not that familiar with JUnit.

By the way, I found another bug regarding exceptions inside `{ message }`. To enable them, I removed the `hasMessage` flag and instead always add messages (seems like the token includes them automagically). Line breaks aren't escaped for JSON, we have to do this ourselves. But I think we can replace all `\n` with `\\n` after building the JSON object string. Gonna give it a try today.

pmwmedia commented on tinylog-impl/src/main/java/org/tinylog/writers/JsonWriter.java:
```diff
@@ -0,0 +1,268 @@
+		if (hasThread) {
+			writeProperty(threadToken, logEntry, builder, hasMethod);
+		}
+		return String.format(JSON_OBJECT, builder.toString());
```
You could build the entire JSON object directly in the target string builder. It would result in an significant performance boost.

proohit replied:
Fixed in https://github.com/tinylog-org/tinylog/pull/195/commits/5322e7bbaaa6040dfad72f601cac39c19db08998

pmwmedia commented:
> I figured out most things like checkstyle, spotbugs etc with your requirement on `mvn verify`. I think we should definitely add a CONTRIBUTING.md guideline. I had some problems with other packages like tinylog-api, but I ignored them as I didn't change anything there.

I totally I agree with you. I have just created an issue for creating a CONTRIBUTING.md guideline (#196).

> If you're curious, here's a link: DirisSlave discord bot.

I'm always happy seeing tinylog used in new projects :)

pmwmedia commented:
After my code review, I got an idea to make the JSON writer more flexible. Currently, the output JSON field names and values are hard coded. Instead, they could be configurable in the tinylog.properties like for the JDBC writer ([documentation](https://tinylog.org/v2/configuration/#jdbc-writer), [source code](https://github.com/tinylog-org/tinylog/blob/baf3306a22e6a40aafddd7c70d76d542dc562465/tinylog-impl/src/main/java/org/tinylog/writers/JdbcWriter.java#L445)).

An example configuration could look like this:

```properties
writer                 = json
writer.level           = debug              # optional
writer.file            = log.txt            # required, absolute or relative path
writer.charset         = UTF-8              # optional
writer.append          = true               # optional, default: false
writer.buffered        = true               # optional, default: false
writer.field.timestamp = date               # short for {date}
writer.field.message   = message            # short for {message}
```

If you take a look into [JdbcWriter.createTokens()](https://github.com/tinylog-org/tinylog/blob/baf3306a22e6a40aafddd7c70d76d542dc562465/tinylog-impl/src/main/java/org/tinylog/writers/JdbcWriter.java#L445), you will see that it took for me less than 10 lines to build tokens from the configuration dynamically.

Please let me know what you think :)

proohit committed:
#194 - Merge Match elements with same bug

proohit committed:
#194 - Replaced manual token assembly

proohit committed:
#194 - First attempt on escaping line breaks

proohit committed:
#194 - Added configurable fields and formats

proohit commented:
Thanks for your awesome review. I'm working on your changes rn.

I really didn't use tools you already provided ;). The tokens do take away a lot of work!

Please take a look in the changes and review if it matches your expectations.

As for tests, I need some more time. I will add them after you approve the implementation

PS: The pipeline got me there with lambda expressions and diamond operators :p

proohit committed:
#194 - Removed diamond operator for java v1.6

proohit committed:
#194 - Removed lambda expression for java v1.6

proohit committed:
#194 - Removed diamond operator for java v1.6

proohit committed:
#194 - Fixed first linebreak bug in exception msg

proohit commented:
I figured out how to handle the line breaks in exception messages. Now, when there is a line seperator, it gets escaped correctly :)

```json
{
    "message": "Hello World!: java.lang.Exception: error message\nat org.tinylog.writers.JsonWriterTest.unappendingWriting(JsonWriterTest.java:57)\nat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)\nat java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)\nat java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\nat java.base/java.lang.reflect.Method.invoke(Method.java:564)\nat...",
    "level": "TRACE"
}
```

pmwmedia commented:
Thank you for your changes! The code looks really clear :) I will review it tomorrow.

proohit committed:
#194 - Added sepcial character escapes

proohit committed:
#194 - Added more tests

pmwmedia commented on tinylog-impl/src/main/java/org/tinylog/writers/JsonWriter.java:
```diff
@@ -0,0 +1,256 @@
+
+		escapeCharacter("\\", "\\\\");
+		escapeCharacter(NEW_LINE, "\\n");
+		escapeCharacter("\t", "\\t");
```
I think we need also to escape quotation marks: `"`

proohit replied:
This one is a bit tricky. At this point, all tokens have rendered the properties, including quotation marks like those in `"message":"asdf"`. Replacing all quotation marks would result in escaping the ones in the properties too, like `\"message\":\"\asdf\"`. Do you have an idea on how to escape only those in the content part of a property?

pmwmedia replied:
You could store all tokens in a map with field name as key and token as value (exactly as read from the configuration without the JSON extras). Then, you can write the field name with all JSON extras to the string builder while outputting a log entry and only escape characters for the real value as rendered by its token.

proohit replied:
Great idea! Gonna try it out tomorrow:)

proohit replied:
I added an implementation for escaping more special characters, though I had to use another StringBuilder. Performance shouldn't be too worse :confused:

pmwmedia commented on tinylog-impl/src/main/java/org/tinylog/writers/JsonWriter.java:
```diff
@@ -0,0 +1,256 @@
+	}
+
+	@Override
+	public Collection<LogEntryValue> getRequiredLogEntryValues() {
```
For me, it would be easier to find methods if they are ordered my visibility. Firstly, all public methods, and all private methods at the end of the class.

proohit replied:
Fixed in https://github.com/tinylog-org/tinylog/pull/195/commits/784dbde0077f730872e7550ea33778c142f7293f

pmwmedia commented:
Just finished the code review. I think the JSON writer is very close to be ready for merging :) I'm going to test the writer in my example project now.

proohit committed:
#194 - Improved escape workflow

proohit committed:
-#194 - Rearrange methods

proohit committed:
#194 - Improved token assembly performance

pmwmedia commented:
While testing on my local Windows laptop, I found some issues:
- Only the first byte of `NEW_LINE` is written in `postprocessFile()`
- An `IOException` is thrown when calling `inputChannel.map()` in `preprocessFile()`
- Invalid trailing comma for the last log entry
- JSON is not formatted with indention and new lines
- The writer is not thread-safe

I have fixed all issues but the last one. Especially, fixing the `IOException` was a bit tricky. Finally, I have created a new `ByteArrayWriter` for `RandomAccessFiles` and removed the `FileChannel`. There is a limitation in my refactored version that the `[` character can only be found in the last 1024 bytes.

My updated version of the `JsonWriter` works on my local Windows laptop, but needs a critical review and is still not thread-safe. However, it is already 1am here and have to continue tomorrow. Nevertheless, I wanted to share my current version.

```java
/*
 * Copyright 2021 Martin Winandy
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

package org.tinylog.writers.raw;

import java.io.IOException;
import java.io.RandomAccessFile;

/**
 * Wrapper for using any {@link RandomAccessFile} as writer.
 */
public final class RandomAccessFileWriter implements ByteArrayWriter {

	private final RandomAccessFile file;

	/**
	 * @param file
	 *            Underlying random access file
	 */
	public RandomAccessFileWriter(final RandomAccessFile file) {
		this.file = file;
	}

	@Override
	public void write(final byte[] data, final int length) throws IOException {
		file.write(data, 0, length);
	}

	@Override
	public void flush() {
	}

	@Override
	public void close() throws IOException {
		file.close();
	}

}
```

```java
/*
 * Copyright 2021 Direnc Timur
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

package org.tinylog.writers;

import java.io.File;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.nio.charset.Charset;
import java.util.Collection;
import java.util.Collections;
import java.util.EnumSet;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.Map.Entry;

import org.tinylog.Level;
import org.tinylog.core.LogEntry;
import org.tinylog.core.LogEntryValue;
import org.tinylog.pattern.FormatPatternParser;
import org.tinylog.pattern.Token;
import org.tinylog.provider.InternalLogger;
import org.tinylog.writers.raw.BufferedWriterDecorator;
import org.tinylog.writers.raw.ByteArrayWriter;
import org.tinylog.writers.raw.RandomAccessFileWriter;
import org.tinylog.writers.raw.SynchronizedWriterDecorator;

/**
 * Writer for outputting log entries to a log file in JSON format. Already
 * existing files can be continued.
 */
public final class JsonWriter implements Writer {

	private static final String NEW_LINE = System.getProperty("line.separator");
	private static final int BUFFER_SIZE = 1024;
	private static final String FIELD_PREFIX = "field.";

	private final Charset charset;
	private ByteArrayWriter writer;
	private boolean firstEntry;

	private StringBuilder builder;
	private final Map<String, Token> jsonProperties;

	private final byte[] newLineBytes;
	private final byte[] bracketOpenBytes;
	private final byte[] bracketCloseBytes;

	/**
	 * @throws IOException              File not found or couldn't access file
	 * @throws IllegalArgumentException Log file is not defined in configuration
	 */
	public JsonWriter() throws IOException {
		this(Collections.<String, String>emptyMap());
	}

	/**
	 * @param properties Configuration for writer
	 * @throws IOException              File not found or couldn't access file
	 * @throws IllegalArgumentException Log file is not defined in configuration
	 */
	public JsonWriter(final Map<String, String> properties) throws IOException {
		jsonProperties = createTokens(properties);
		charset = getCharset(properties);
		newLineBytes = NEW_LINE.getBytes(charset);
		bracketOpenBytes = "[".getBytes(charset);
		bracketCloseBytes = "]".getBytes(charset);

		String fileName = getFileName(properties);
		File file = new File(fileName).getAbsoluteFile();
		file.getParentFile().mkdirs();

		boolean append = Boolean.parseBoolean(properties.get("append"));

		RandomAccessFile randomAccessFile = new RandomAccessFile(file, "rw");

		if (append) {
			randomAccessFile.seek(randomAccessFile.length());
		} else {
			randomAccessFile.setLength(0);
		}

		writer = new RandomAccessFileWriter(randomAccessFile);

		boolean buffered = Boolean.parseBoolean(properties.get("buffered"));
		if (buffered) {
			writer = new BufferedWriterDecorator(writer);
		}

		boolean writingThread = Boolean.parseBoolean(properties.get("writingthread"));
		if (!writingThread) {
			writer = new SynchronizedWriterDecorator(writer, randomAccessFile);
		}

		preprocessFile(randomAccessFile, append);
	}

	@Override
	public void write(final LogEntry logEntry) throws IOException {
		if (builder == null) {
			builder = new StringBuilder();
		} else {
			builder.setLength(0);
		}
		addJsonObject(logEntry);
		writer.write(builder.toString().getBytes(charset), builder.length());
	}

	@Override
	public void flush() throws IOException {
		writer.flush();
	}

	@Override
	public void close() throws IOException {
		postprocessFile();
		writer.close();
	}

	@Override
	public Collection<LogEntryValue> getRequiredLogEntryValues() {
		Collection<LogEntryValue> values = EnumSet.noneOf(LogEntryValue.class);
		for (Token token : jsonProperties.values()) {
			values.addAll(token.getRequiredLogEntryValues());
		}
		return values;
	}

	/**
	 * Prepares and adds a Json Object. Also escapes special characters.
	 *
	 * @param logEntry LogEntry with information for token
	 */
	private void addJsonObject(final LogEntry logEntry) {
		if (!firstEntry) {
			builder.append(',');
		}

		builder.append(NEW_LINE);
		builder.append('\t');
		builder.append("{");

		Token[] tokenEntries = jsonProperties.values().toArray(new Token[0]);
		String[] fields = jsonProperties.keySet().toArray(new String[0]);

		StringBuilder tokenStringBuilder = new StringBuilder();
		for (int i = 0; i < tokenEntries.length; i++) {
			builder.append("\"").append(fields[i]).append("\":\"");

			Token token = tokenEntries[i];
			tokenStringBuilder.setLength(0);
			token.render(logEntry, tokenStringBuilder);

			escapeCharacter("\\", "\\\\", tokenStringBuilder);
			escapeCharacter("\"", "\\\"", tokenStringBuilder);
			escapeCharacter(NEW_LINE, "\\n", tokenStringBuilder);
			escapeCharacter("\t", "\\t", tokenStringBuilder);
			escapeCharacter("\b", "\\b", tokenStringBuilder);
			escapeCharacter("\f", "\\f", tokenStringBuilder);
			escapeCharacter("\n", "\\n", tokenStringBuilder);
			escapeCharacter("\r", "\\r", tokenStringBuilder);

			builder.append(tokenStringBuilder.toString()).append("\"");

			if (i + 1 < jsonProperties.size()) {
				builder.append(", ");
			}
		}
		builder.append("}");

		if (firstEntry) {
			firstEntry = false;
		}
	}

	private void escapeCharacter(final String character, final String escapeWith, final StringBuilder stringBuilder) {
		int index = stringBuilder.indexOf(character);
		while (index != -1) {
			stringBuilder.replace(index, index + character.length(), escapeWith);
			index += escapeWith.length();
			index = stringBuilder.indexOf(character, index);
		}
	}

	private Charset getCharset(final Map<String, String> properties) {
		String charsetName = properties.get("charset");
		try {
			return charsetName == null ? Charset.defaultCharset() : Charset.forName(charsetName);
		} catch (IllegalArgumentException ex) {
			InternalLogger.log(Level.ERROR, "Invalid charset: " + charsetName);
			return Charset.defaultCharset();
		}
	}

	private String getFileName(final Map<String, String> properties) {
		String fileName = properties.get("file");
		if (fileName == null) {
			throw new IllegalArgumentException("File name is missing for file writer");
		}
		return fileName;
	}

	/**
	 * Preprocesses the JSON file. If append mode is on, deletes the closing bracket
	 * and adds a comma instead. If it's a new file, appends an opening bracket.
	 *
	 *
	 * @param file   Log file
	 * @param append Append Mode on or off
	 * @throws IOException              Error reading or writing file
	 * @throws IllegalArgumentException Invalid file format
	 */
	private void preprocessFile(RandomAccessFile file, final boolean append) throws IOException, IllegalArgumentException {
		if (append && file.length() > 0) {
			long sizeToTruncate = 0;
			long currentPosition = file.getFilePointer();
			boolean foundClosingBracket = false;

			int numberOfBytes = (int) Math.min(currentPosition, BUFFER_SIZE);
			byte[] bytes = new byte[numberOfBytes];
			file.seek(currentPosition - numberOfBytes);
			file.read(bytes);

			for (int i = bytes.length - 1; i >= 0; i--, currentPosition--) {
				sizeToTruncate += 1;
				if (!foundClosingBracket && bytes[i] == ']') {
					foundClosingBracket = true;
				} else if (foundClosingBracket && bytes[i] != '\n' && bytes[i] != '\r') {
					sizeToTruncate -= 1;
					break;
				}
			}

			if (!foundClosingBracket) {
				throw new IllegalArgumentException(
						"Invalid JSON file. The file is missing a closing bracket for the array.");
			}

			long newFileSize = file.length() - sizeToTruncate;
			file.setLength(newFileSize);
			file.seek(newFileSize);
		}

		if (file.length() == 0) {
			writer.write(bracketOpenBytes, bracketOpenBytes.length);
		}

		file.seek(file.length() - 1);
		firstEntry = file.read() == '[';
	}

	/**
	 * Postprocesses the JSON file. Attempts to delete the trailing comma and
	 * appends closing bracket which were handled in
	 * {@link #preprocessFile(RandomAccessFile, boolean)}
	 *
	 * @throws IOException Error writing to file
	 */
	private void postprocessFile() throws IOException {
		writer.write(newLineBytes, newLineBytes.length);
		writer.write(bracketCloseBytes, bracketCloseBytes.length);
	}

	private static Map<String, Token> createTokens(final Map<String, String> properties) {
		FormatPatternParser parser = new FormatPatternParser(properties.get("exception"));

		Map<String, Token> tokens = new HashMap<String, Token>();
		for (Entry<String, String> entry : properties.entrySet()) {
			if (entry.getKey().toLowerCase(Locale.ROOT).startsWith(FIELD_PREFIX)) {
				tokens.put(entry.getKey().substring(FIELD_PREFIX.length()), parser.parse(entry.getValue()));
			}
		}
		return tokens;
	}

}
```

proohit commented:
Thanks for your changes!

Could you maybe clarify what exception was thrown with map?

Hmm, after some quick research, many people suggest to map the data into memory and managing thread safety from there. According to https://docs.oracle.com/javase/7/docs/api/java/io/RandomAccessFile.html , there is a rws (read/write synchronized) access mode on RandomAccessFile, maybe we use it.

pmwmedia commented:
> Could you maybe clarify what exception was thrown with map?

I have received an IOException with a German message: java.io.IOException: Der Vorgang ist bei einer Datei mit einem geöffneten Bereich, der einem Benutzer zugeordnet ist, nicht anwendbar. It means something like "The operation is not applicable to a file with an open area assigned to a user". I assume that Windows has issues with using the same file by multiple streams. Therefore, I switched to RandomAccessFile.

> Hmm, after some quick research, many people suggest to map the data into memory and managing thread safety from there. According to https://docs.oracle.com/javase/7/docs/api/java/io/RandomAccessFile.html , there is a rws (read/write synchronized) access mode on RandomAccessFile, maybe we use it.

According to the Javadoc, "rws" requires at least one more low-level I/O operation in opposite to "rw". Therefore, I think that using the `SynchronizedWriterDecorator` is sufficient for our use case.

pmwmedia commented:
I hope that I could fix all thread-safety issues. About the trailing comma, I switched back to your solution.

Please review my code critically and feel free to change, rearrange, and improve it. 

```java
/*
 * Copyright 2021 Direnc Timur
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */

package org.tinylog.writers;

import java.io.File;
import java.io.IOException;
import java.io.RandomAccessFile;
import java.nio.charset.Charset;
import java.util.Collection;
import java.util.Collections;
import java.util.EnumSet;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.Map.Entry;

import org.tinylog.Level;
import org.tinylog.core.LogEntry;
import org.tinylog.core.LogEntryValue;
import org.tinylog.pattern.FormatPatternParser;
import org.tinylog.pattern.Token;
import org.tinylog.provider.InternalLogger;
import org.tinylog.writers.raw.BufferedWriterDecorator;
import org.tinylog.writers.raw.ByteArrayWriter;
import org.tinylog.writers.raw.RandomAccessFileWriter;
import org.tinylog.writers.raw.SynchronizedWriterDecorator;

/**
 * Writer for outputting log entries to a log file in JSON format. Already
 * existing files can be continued.
 */
public final class JsonWriter implements Writer {

	private static final String NEW_LINE = System.getProperty("line.separator");
	private static final int BUFFER_SIZE = 1024;
	private static final String FIELD_PREFIX = "field.";

	private final Charset charset;
	private final RandomAccessFile randomAccessFile;
	private ByteArrayWriter writer;

	private StringBuilder builder;
	private final Map<String, Token> jsonProperties;

	private final byte[] newLineBytes;
	private final byte[] commaBytes;
	private final byte[] bracketOpenBytes;
	private final byte[] bracketCloseBytes;

	/**
	 * @throws IOException              File not found or couldn't access file
	 * @throws IllegalArgumentException Log file is not defined in configuration
	 */
	public JsonWriter() throws IOException {
		this(Collections.<String, String>emptyMap());
	}

	/**
	 * @param properties Configuration for writer
	 * @throws IOException              File not found or couldn't access file
	 * @throws IllegalArgumentException Log file is not defined in configuration
	 */
	public JsonWriter(final Map<String, String> properties) throws IOException {
		jsonProperties = createTokens(properties);
		charset = getCharset(properties);
		newLineBytes = NEW_LINE.getBytes(charset);
		commaBytes = ",".getBytes(charset);
		bracketOpenBytes = "[".getBytes(charset);
		bracketCloseBytes = "]".getBytes(charset);

		String fileName = getFileName(properties);
		File file = new File(fileName).getAbsoluteFile();
		file.getParentFile().mkdirs();

		boolean append = Boolean.parseBoolean(properties.get("append"));

		randomAccessFile = new RandomAccessFile(file, "rw");
		writer = new RandomAccessFileWriter(randomAccessFile);

		boolean buffered = Boolean.parseBoolean(properties.get("buffered"));
		if (buffered) {
			writer = new BufferedWriterDecorator(writer);
		}

		boolean writingThread = Boolean.parseBoolean(properties.get("writingthread"));
		if (writingThread) {
			builder = new StringBuilder();
		} else {
			writer = new SynchronizedWriterDecorator(writer, randomAccessFile);
		}

		preProcessFile(append);
	}

	@Override
	public void write(final LogEntry logEntry) throws IOException {
		StringBuilder builder;
		if (this.builder == null) {
			builder = new StringBuilder();
		} else {
			builder = this.builder;
			builder.setLength(0);
		}

		addJsonObject(logEntry, builder);
		writer.write(builder.toString().getBytes(charset), builder.length());
	}

	@Override
	public void flush() throws IOException {
		writer.flush();
	}

	@Override
	public void close() throws IOException {
		writer.flush();
		postProcessFile();
		writer.close();
	}

	@Override
	public Collection<LogEntryValue> getRequiredLogEntryValues() {
		Collection<LogEntryValue> values = EnumSet.noneOf(LogEntryValue.class);
		for (Token token : jsonProperties.values()) {
			values.addAll(token.getRequiredLogEntryValues());
		}
		return values;
	}

	/**
	 * Prepares and adds a Json Object. Also escapes special characters.
	 *
	 * @param logEntry LogEntry with information for token
	 * @param builder  Target for the created the JSON object
	 */
	private void addJsonObject(final LogEntry logEntry, final StringBuilder builder) {
		builder.append(NEW_LINE);
		builder.append('\t');
		builder.append("{");

		Token[] tokenEntries = jsonProperties.values().toArray(new Token[0]);
		String[] fields = jsonProperties.keySet().toArray(new String[0]);

		StringBuilder tokenStringBuilder = new StringBuilder();
		for (int i = 0; i < tokenEntries.length; i++) {
			builder.append("\"").append(fields[i]).append("\":\"");

			Token token = tokenEntries[i];
			tokenStringBuilder.setLength(0);
			token.render(logEntry, tokenStringBuilder);

			escapeCharacter("\\", "\\\\", tokenStringBuilder);
			escapeCharacter("\"", "\\\"", tokenStringBuilder);
			escapeCharacter(NEW_LINE, "\\n", tokenStringBuilder);
			escapeCharacter("\t", "\\t", tokenStringBuilder);
			escapeCharacter("\b", "\\b", tokenStringBuilder);
			escapeCharacter("\f", "\\f", tokenStringBuilder);
			escapeCharacter("\n", "\\n", tokenStringBuilder);
			escapeCharacter("\r", "\\r", tokenStringBuilder);

			builder.append(tokenStringBuilder.toString()).append("\"");

			if (i + 1 < jsonProperties.size()) {
				builder.append(", ");
			}
		}
		builder.append("},");
	}

	private void escapeCharacter(final String character, final String escapeWith, final StringBuilder stringBuilder) {
		int index = stringBuilder.indexOf(character);
		while (index != -1) {
			stringBuilder.replace(index, index + character.length(), escapeWith);
			index += escapeWith.length();
			index = stringBuilder.indexOf(character, index);
		}
	}

	private Charset getCharset(final Map<String, String> properties) {
		String charsetName = properties.get("charset");
		try {
			return charsetName == null ? Charset.defaultCharset() : Charset.forName(charsetName);
		} catch (IllegalArgumentException ex) {
			InternalLogger.log(Level.ERROR, "Invalid charset: " + charsetName);
			return Charset.defaultCharset();
		}
	}

	private String getFileName(final Map<String, String> properties) {
		String fileName = properties.get("file");
		if (fileName == null) {
			throw new IllegalArgumentException("File name is missing for file writer");
		}
		return fileName;
	}

	/**
	 * Pre-processes the JSON file. If append mode is on, deletes the closing bracket
	 * and adds a comma instead. If it's a new file, appends an opening bracket.
	 *
	 *
	 * @param append Append Mode on or off
	 * @throws IOException              Error reading or writing file
	 * @throws IllegalArgumentException Invalid file format
	 */
	private void preProcessFile(final boolean append) throws IOException, IllegalArgumentException {
		if (append && randomAccessFile.length() > 0) {
			long sizeToTruncate = 0;
			boolean foundClosingBracket = false;

			byte[] bytes = new byte[BUFFER_SIZE];
			randomAccessFile.seek(Math.max(0, randomAccessFile.length() - BUFFER_SIZE));
			int numberOfBytes = randomAccessFile.read(bytes);

			for (int i = numberOfBytes - 1; i >= 0; i--) {
				byte letter = bytes[i];
				sizeToTruncate += 1;

				if (!foundClosingBracket && letter == ']') {
					foundClosingBracket = true;
				} else if (foundClosingBracket && letter != '\n' && letter != '\r') {
					sizeToTruncate -= 1;
					break;
				}
			}

			if (!foundClosingBracket) {
				throw new IllegalArgumentException(
						"Invalid JSON file. The file is missing a closing bracket for the array.");
			}

			long newFileSize = randomAccessFile.length() - sizeToTruncate;
			randomAccessFile.setLength(newFileSize);
			randomAccessFile.seek(newFileSize);
		} else {
			randomAccessFile.setLength(0);
		}

		if (randomAccessFile.length() == 0) {
			writer.write(bracketOpenBytes, bracketOpenBytes.length);
		} else {
			randomAccessFile.seek(randomAccessFile.length() - 1);
			if (randomAccessFile.read() == '}') {
				writer.write(commaBytes, commaBytes.length);
			}
		}
	}

	/**
	 * Post-processes the JSON file. Attempts to delete the trailing comma and
	 * appends closing bracket which were handled in {@link #preProcessFile(boolean)}.
	 *
	 * @throws IOException Error writing to file
	 */
	private void postProcessFile() throws IOException {
		if (randomAccessFile.length() > 0) {
			randomAccessFile.seek(randomAccessFile.length() - 1);
			if (randomAccessFile.read() == ',') {
				randomAccessFile.setLength(randomAccessFile.length() - 1);
			}
		}

		writer.write(newLineBytes, newLineBytes.length);
		writer.write(bracketCloseBytes, bracketCloseBytes.length);
	}

	private static Map<String, Token> createTokens(final Map<String, String> properties) {
		FormatPatternParser parser = new FormatPatternParser(properties.get("exception"));

		Map<String, Token> tokens = new HashMap<String, Token>();
		for (Entry<String, String> entry : properties.entrySet()) {
			if (entry.getKey().toLowerCase(Locale.ROOT).startsWith(FIELD_PREFIX)) {
				tokens.put(entry.getKey().substring(FIELD_PREFIX.length()), parser.parse(entry.getValue()));
			}
		}
		return tokens;
	}

}
```

proohit committed:
#194 - Refactored JsonWriter, use RandomAccessFile

These changes came in authored by @pmwmedia as suggested in #195 and refactored by me

proohit committed:
#194 - Added tests

proohit committed:
#194 - Added JSON formatting, fixed linting

proohit commented:
Hi, I added your changes and made some refactoring. I also added a test to the RandomAccessFileWriter. 

I included JSON formatting with tabs and line breaks, although I think that JSON is format-less and we could omit formatting. That's just my opinion though. Please have a look at the changes!

EDIT: I also updated the proposed documentation entry inside the initial PR description :)

pmwmedia commented:
@proohit: Thank you very much for your changes! The code looks perfect :) I only miss a test for `getRequiredLogEntryValues()`. Could you add a test case? Afterwards I'm going to merge your pull request.

proohit committed:
#194 - Added getRequiredLogEntryValues test

proohit commented:
Hey, I've  added the test.

Looking forward to see the JsonWriter in production :)

pmwmedia commented:
@proohit: I'm happy to merge your pull request! It is always interesting how many edge cases have to be considered. Thank you for bringing the original concept into a solid and flexible solution with good test coverage.
